File: simple-message/src/main/java/ru/starksoft/simplemessage/SimpleMessage.java
package ru.starksoft.simplemessage;

import android.app.Activity;
import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.OnLifecycleEvent;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import org.jetbrains.annotations.NotNull;

import static ru.starksoft.simplemessage.util.Util.checkNotNull;

@SuppressWarnings("WeakerAccess")
public final class SimpleMessage {

	private static final String TAG = "SimpleMessage";

	@NonNull private final Activity activity;

	private final StatusBarMessageLayout view;

	private SimpleMessage(@NonNull Activity activity, @NonNull MessageRecord messageRecord) {
		this.activity = checkNotNull(activity);

		view = new StatusBarMessageLayout(activity, messageRecord);

		observeLifecycle();
	}

	@NonNull
	public static Builder create(@NonNull Activity activity, @NonNull CharSequence message) {
		return new Builder(activity, message, MessageTemplate.NONE);
	}

	@NonNull
	public static Builder create(@NonNull Activity activity, @NonNull CharSequence message, @MessageTemplate int template) {
		return new Builder(activity, message, template);
	}

	private void observeLifecycle() {
		if (activity instanceof AppCompatActivity) {
			((AppCompatActivity) activity).getLifecycle().addObserver(new LifecycleObserver() {

				@OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
				void destroy() {
					SimpleMessageManager.getInstance().destroy();
					Log.d(TAG, "destroy() called");
				}
			});
		}
	}

	public void show() {
		view.show();
	}

	public void hide() {
		view.hide();
	}

	public static final class Builder {

		private final MessageRecord messageRecord;
		@NonNull private final Activity activity;

		public Builder(@NonNull Activity activity, @NonNull CharSequence message, @MessageTemplate int template) {
			this.activity = activity;
			messageRecord = getMessageRecordByTemplate(message, template);
			messageRecord.setMessage(message);
		}

		@NonNull
		private MessageRecord getMessageRecordByTemplate(@NonNull CharSequence message, @MessageTemplate int template) {
			switch (template) {
				case MessageTemplate.ERROR_PERSISTENT:
					return MessageRecord.createErrorPersistent(message);

				case MessageTemplate.ERROR:
					return MessageRecord.createError(message);

				case MessageTemplate.NORMAL_PROGRESS:
					return MessageRecord.createNormalWithProgress(message);

				case MessageTemplate.NORMAL:
					return MessageRecord.createNormal(message);

				case MessageTemplate.SUCCESS:
					return MessageRecord.createSuccess(message);

				case MessageTemplate.NONE:
				default:
					return new MessageRecord();
			}
		}

		@NotNull
		public Builder textColor(@ColorInt int color) {
			messageRecord.setTextColor(color);
			return this;
		}

		@NotNull
		public Builder backgroundColor(@ColorInt int color) {
			messageRecord.setBackgroundColor(color);
			return this;
		}

		@NotNull
		public Builder progress(boolean progress) {
			messageRecord.setProgress(progress);
			return this;
		}

		@NotNull
		public Builder persistent(boolean persistent) {
			messageRecord.setPersistent(persistent);
			return this;
		}

		@NonNull
		public SimpleMessage build() {
			return new SimpleMessage(activity, messageRecord);
		}

		public void show() {
			build().show();
		}
	}
}

File: simple-message/src/main/java/ru/starksoft/simplemessage/SimpleMessage.kt
package ru.starksoft.simplemessage

import android.app.Activity
import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.OnLifecycleEvent
import android.support.annotation.ColorInt
import android.support.v7.app.AppCompatActivity
import android.util.Log

class SimpleMessage private constructor(activity: Activity, messageData: MessageData) {

	private val activity: Activity = checkNotNull(activity)

	private val view: Message = getMessageImplByType(activity, messageData)

	init {
		observeLifecycle()
	}

	private fun observeLifecycle() {
		if (activity is AppCompatActivity) {
			activity.lifecycle.addObserver(object : LifecycleObserver {
				@OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
				fun destroy() {
					SimpleMessageManager.instance.destroy()
					Log.d(TAG, "destroy() called")
				}
			})
		}
	}

	fun show() {
		view.show()
	}

	fun hide() {
		view.hide()
	}

	private fun getMessageImplByType(activity: Activity, messageData: MessageData): Message {
		return when (messageData.messageType) {
			MessageType.STATUS_BAR -> StatusBarMessageLayout(activity, messageData)

			MessageType.TOOL_BAR -> ToolBarMessageLayout(activity, messageData)
		}
	}

	@Suppress("MemberVisibilityCanBePrivate")
	class Builder(private val activity: Activity, message: CharSequence, template: MessageTemplate?) {

		private val messageData: MessageData

		init {
			messageData = getMessageRecordByTemplate(message, template)
			messageData.message = message
		}

		private fun getMessageRecordByTemplate(message: CharSequence, template: MessageTemplate?): MessageData {
			return when (template) {
				MessageTemplate.ERROR_PERSISTENT -> MessageData.createErrorPersistent(message)

				MessageTemplate.ERROR -> MessageData.createError(message)

				MessageTemplate.NORMAL_PROGRESS -> MessageData.createNormalWithProgress(message)

				MessageTemplate.NORMAL -> MessageData.createNormal(message)

				MessageTemplate.SUCCESS -> MessageData.createSuccess(message)

				else -> MessageData()
			}
		}

		fun textColor(@ColorInt color: Int): Builder {
			messageData.textColor = color
			return this
		}

		fun backgroundColor(@ColorInt color: Int): Builder {
			messageData.backgroundColor = color
			return this
		}

		fun progress(progress: Boolean): Builder {
			messageData.isProgress = progress
			return this
		}

		fun persistent(persistent: Boolean): Builder {
			messageData.isPersistent = persistent
			return this
		}

		fun messageType(messageType: MessageType): Builder {
			messageData.messageType = messageType
			return this
		}

		fun build(): SimpleMessage {
			return SimpleMessage(activity, messageData)
		}

		fun show() {
			build().show()
		}
	}

	companion object {

		private const val TAG = "SimpleMessage"

		@JvmStatic
		fun create(activity: Activity, message: CharSequence): Builder {
			return create(activity, message, null)
		}

		@JvmStatic
		fun create(activity: Activity, message: CharSequence, template: MessageTemplate?): Builder {
			return Builder(activity, message, template)
		}

		fun hide() {
			SimpleMessageManager.instance.hide()
		}
	}
}
